src/main.rs

```diff
@@ -470,55 +526,189 @@ fn verify_event_attendance_and_get_event_name(spec: &CredentialSpec) -> Result<&
     }
 }
 
+// TODO: Create and manage admins instead of using canister controllers.
+async fn is_admin(id: Principal) -> bool {
+    is_controller(&id)
+}
+
 #[update]
 #[candid_method]
-fn register_early_adopter(
-    request: RegisterRequest,
-) -> Result<EarlyAdopterResponse, EarlyAdopterError> {
+async fn list_events() -> Result<ListEventsResponse, RegisterError> {
+    let user_id = caller();
+    let is_controller = is_admin(user_id).await;
+    EVENTS.with_borrow(|events| {
+        let events: Vec<EventData> = events
+            .iter()
+            .map(|(event_name, data)| EventData {
+                created_timestamp_s: data.created_timestamp_s.clone(),
+                event_name: event_name.clone(),
+                code: if is_controller { Some(data.code) } else { None },
+            })
+            .collect();
+        Ok(ListEventsResponse { events })
+    })
+}
+
+// Returns a random string of length 24
+async fn generate_random_code() -> Option<String> {
```

Comment:
Maybe add more description about the code being generated (and rename the function to be more specific)?

Reply:
Done!